### ISSUE TITLE
XP Tracker is miscalculating xp gained / hour

### ISSUE THREAD

**Issue by Gehbrial** [runelite/runelite#10099]
**Describe the bug**
XP Tracker is miscalculating xp gained / hour. The XP/hour keeps going up instead of readjusting itself.

**To Reproduce**
I *think* it worked fine at first but started acting up once I clicked on "Reset All". I'm not sure.

**Expected behavior**
It should give you an estimate of the xp gained per hour instead of always going up.

**Screenshots**
![xp-tracker](https://user-images.githubusercontent.com/9321998/67131581-37519e00-f1d3-11e9-8bb8-eeb465a9262a.PNG)

**Environment (please complete the following information):**
 - OS: Windows 10
 - RuneLite version: 1.5.36
 - Launcher version: 1.60

**Comment by NathanielPather** [comment]
Do you remember if the total xp per hour number counted down at all?

**Comment by critzd** [comment]
How long did the incorrect XP/Hour continue to occur? Did it stop after 60 seconds of using a skill?

**Comment by critzd** [comment]
After looking into it further, it seems that for the first 60 seconds that a skill is active, any time relative calculations for the xp tracker will default to 60s to prevent any crazy xp/hour or other calculations from appearing. There's always the option to reduce that grace period, but as it stands right now this seems intentional.

I haven't been able to reproduce anything outside of that 60 second window, but if that's what's occurring to you please mention it as that would mean a different issue may be the cause.

**Comment by Gehbrial** [comment]
It stopped when I closed the client several hours later. I was afking magic in NMZ if it is of any help. I'm sorry I'm unable to provide steps on how to reproduce it.

**Comment by critzd** [comment]
I'll do some further digging, thanks for the info!

**Comment by whartd** [comment]
I have no idea how to reproduce this, but I've encountered it before. It's been months since I've seen it though, and it was rare

**Comment by RedSparr0w** [comment]
Commented on #15079 but just realised that issue was closed, so figured I'd post it here too.

Just had this same bug.
![image](https://user-images.githubusercontent.com/7288322/206822295-5ade2d1e-a42c-44cb-8ce9-c9a1d9e33ae9.png)

Seems it was caused by my computers date going backwards (even after reseting the tracking, it would still just go up and up)

**Comment by Nightfirecat** [comment]
This is caused because the plugin uses `System.currentTimeMillis()` (which can give values from the past if the system time changes between calls) rather than the monotonic `System.nanoTime()`.

**Comment by robotboy655** [comment]
There seems to be a systemic issue with any "per hour" counter where it appears that the "per hour" value is calculated using the entire time counted since the last reset, instead of doing a rolling average that would show the actual "per hour" value. (i.e. discarding data older than an hour or two) This means the "per hour" counters will never be accurate, making them kind of useless most of the time.

Asking players to constantly reset the XP/Hr counters in XP tracker is not a real solution.

The system could be updated to have both values - the long term counter since last reset, and a rolling average with the real value if preserving the existing behavior is useful to some players. One of the values could be shown in parenthesis in the XP tracker for example.

It makes no sense to show that I am getting 1440xp/hr on skills I have not trained in days.

**Comment by JZomDev** [comment]
@robotboy655 
> It makes no sense to show that I am getting 1440xp/hr on skills I have not trained in days.

In January 2025 RuneLite had an update which default enabled `Save between sessions` in the XP Tracker config. That is why your xp/hr is far below where it should say. If you don't want that feature, turn it off & your issue is not related to the issue this user reported in 2019.

**Comment by robotboy655** [comment]
You are right, it's not related, my bad, I did not see that it's kind of the opposite.

But not saving the XP is again, not a solution.
I do want the XP totals to save, but I do not want the xp/hr counters to not work correctly.
It also does not really fix the issue at all if you play for more than say 2 hours in a session, which is highly likely.

**Comment by JZomDev** [comment]
Consider using XP updater to track your total xp gained in periods to services like wiseoldman or templeosrs. Wise Old Man has a right click option at the top of the XP Tracker's side panel built in. This would help keep track of total xp gained over time periods. 

xp/hr is always going to be contextual to what you're doing IE attaching feathers to arrow shafts vs fletching yew long bows, if you want to get a better xp/hr then you have to reset that stat. If you switch between the two activities in the same hour you'd see the average of the two activities and not two different numbers. I think the approach is fine. I don't know of a way for the plugin to be aware to reset stats in anyway that would cover all basis, making it manual is a good solution.

**Comment by raiyni** [comment]
Set an auto reset time.

**Comment by robotboy655** [comment]
@JZomDev I do not consider "install something else" to be a solution, especially when core behavior can be just fixed.

> xp/hr is always going to be contextual to what you're doing IE attaching feathers to arrow shafts vs fletching yew long bows

That's perfectly normal and not what I am talking about at all. I am specifically talking about the fact that "per hour" counters do not reflect reality after about 2 hours of play time. Mixing actions would show an average and that is totally how it should work.

@raiyni  I guess auto reset time might be a decent enough bandaid for this. I wasn't aware it would only affect the "per hour" xp rates.

**Comment by JZomDev** [comment]
XP Updater is core and not something you need to install. 

Alternatively you can auto pause after a few minutes so when you take a break to do a farm run and resume training, the xp/hr of the old number continues to be averaged with xp continued to be gained rather than just clearing it out.

**Comment by raiyni** [comment]
You can also enable the auto pause feature!!!

**Comment by robotboy655** [comment]
Honestly the developers may want to set up some sane defaults for those 2 options to avoid people like me and others (https://github.com/runelite/runelite/issues/18786 https://github.com/runelite/runelite/issues/19000) from creating issues about this if that's the indented solution.

**Comment by cdfisher** [comment]
Can you maybe open a discussion about this rather than continuing here? As I understand it, what you're asking about is more of just wanting a change in the existing behavior, rather than what this issue was originally opened about (which was actually incorrect behavior).
